src/dmd/initsem.d

```diff
@@ -408,6 +408,10 @@ private extern(C++) final class InitializerSemanticVisitor : Visitor
         if (t.ty == Tsarray)
         {
             uinteger_t edim = (cast(TypeSArray)t).dim.toInteger();
+            if (!i.isAssociativeArray() && i.dim != edim)
+            {
+                deprecation(i.loc, "mismatched array lengths, %llu and %u", edim, i.dim);
```

Comment:
Needs a @@@DEPRECATED comment as described in https://github.com/dlang/DIPs/blob/master/DIPs/DIP1013.md